### PR TITLE
Runtime validation info

### PR DESCRIPTION
**Pull request by j-mueller**
Emulator generates `PendingTx` values at validation time. Now we don't have to have to keep TH expressions around for each mockchain state that needs to be validated in the tests.

### CODE REVIEW

File: core-to-plc/src/Language/Plutus/CoreToPLC/Plugin.hs [hunk 1]
```diff
@@ -75,6 +75,14 @@ instance FromJSON PlcCode where
       Left e  -> fail e
       Right v -> pure v
 
+lifted :: LiftPlc a => a -> PlcCode
```
Comment (michaelpj) [inline core-to-plc/src/Language/Plutus/CoreToPLC/Plugin.hs]: I'm not wild about this, and here's I think why. `PlcCode` is really supposed to be the opaque type produced by the plugin. As such I want to expose as few operations on it as possible, since otherwise it's harder to change.

The upshot of this is that we perhaps shouldn't use it as the type of code on the chain. We might use another type for that (probably isomorphic since it's going to also contain serialized code) but crucially I can change `PlcCode` without affecting that.

Possibly nothing to do here, maybe I'll do something in a follow-up.

File: plutus-use-cases/test/Spec/Crowdfunding.hs [hunk 2]
```diff
@@ -84,13 +76,7 @@ successfulCampaign = checkCFTrace scenario1 $ do
...
     con3 <- contrib3 c 800
     updateAll'
-    setValidationData $ ValidationData $(plutus [| let el = []::[Signature] in
-            $(pendingTxCrowdfunding)
-                11
-                (Signature 1:el)
-                (600, Signature 2 : el)
-                (Just (800, Signature 3 : el))
-            |])
+    addBlocks 10
```
Comment (michaelpj) [inline plutus-use-cases/test/Spec/Crowdfunding.hs]: why is this added?

File: plutus-use-cases/test/Spec/Crowdfunding.hs [hunk 3]
```diff
@@ -147,22 +119,9 @@ canRefund = checkCFTrace scenario1 $ do
...
     con3 <- contrib3 c 800
     updateAll'
-    setValidationData $ ValidationData $(plutus [| let el = []::[Signature] in
-            $(pendingTxCrowdfunding)
-                18
-                (Signature 2 : el)
-                (600, Signature 2 : el)
-                Nothing
-            |])
+    addBlocks 15
```
Comment (michaelpj) [inline plutus-use-cases/test/Spec/Crowdfunding.hs]: ditto

Comment (michaelpj) [reply plutus-use-cases/test/Spec/Crowdfunding.hs]: happens in a few other places too

File: plutus-use-cases/test/Spec/Crowdfunding.hs (hunk 2 above)
Comment (j-mueller) [reply plutus-use-cases/test/Spec/Crowdfunding.hs]: To ensure that enough "time" (blocks) has passed so that we can retrieve the funds or claim a refund. 

Block height is passed to the validation script as part of the `PendingTx` value. Previously we had to encode the desired `PendingTx` value in TH in order to compile it with the plugin (so the block height of the mockchain was different from what we passed to the validator script). Now the emulator uses the actual mockchain height, which (at this stage in the `Trace`) is no large enough for any of the conditions in the validator script to become true. So we have to add a number of empty blocks to simulate the passing of time.

Comment (michaelpj) [reply plutus-use-cases/test/Spec/Crowdfunding.hs]: Perhaps you should add some comments to your mock traces explaining what they're doing :)

File: core-to-plc/src/Language/Plutus/CoreToPLC/Plugin.hs (hunk 1 above)
Comment (j-mueller) [reply core-to-plc/src/Language/Plutus/CoreToPLC/Plugin.hs]: I agree, it makes sense to split up the two uses of `PlcCode`, but changing that should happen in a different PR (adding `lifted` is just a consequence of the fact that `PlcCode` is also used as the on-chain type right now)

The operations we need on `PlcCode` _after_ it has been put on the chain are

* `apply` to other `PlcCode`s and values that can be `lift`ed
* `evaluate`

So maybe we need a `newtype OnChainCode = OnChainCode ByteString` with `apply`, `evaluate`, `lifted` and `PlcCode -> OnChainCode` ?

File: plutus-use-cases/test/Spec/Crowdfunding.hs (hunk 2 above)
Comment (j-mueller) [reply plutus-use-cases/test/Spec/Crowdfunding.hs]: Good idea...

File: core-to-plc/src/Language/Plutus/CoreToPLC/Plugin.hs (hunk 1 above)
Comment (michaelpj) [reply core-to-plc/src/Language/Plutus/CoreToPLC/Plugin.hs]: Exactly! It could probably even be a `SerializedProgram` type in `language-plutus-core`.

**Comment by michaelpj** [comment]
:+1:

**Comment by michaelpj** [comment]
I'll look into the `PlcCode` issue.
